### ISSUE TITLE
Gravity and orientation sensors don't work

### ISSUE THREAD

**Issue by glokuosta** [markusfisch/ShaderEditor#90]
I am on the 2.14 version and I tried importing gravity and orientation samples, and they don't seem to work on my device. I have a Samsung Galaxy C7 with Android 8.0.0

I remember those samples were working before, but I am not sure it was before or after I updated from Android 7 about two months ago. So I downloaded 2.12 from F-Droid and fresh installed that, but they still didn't work. I had an old backup from v2.6.1 and tried that as well, and the uniforms were working on that version. 

Currently I'm on 2.14 and tried a couple of other sensors. Linear, gyro and inclination uniforms doesn't change the output of the shader, but I don't know what they do exactly, I just shook my device and nothing changed. Light and magnetic sensors work, so not all of them are broken. I tried disabling power saving mode, phone orientation, looked into app settings to see if there are some permissions I haven't given, changing sensor delays in the settings, but nothing changed anything.

**Comment by markusfisch** [comment]
Thanks for reporting! I saw this a couple of times on some of my test devices, too. But it was quite rare and didn't just affect Shader Editor but other apps reading the acceleration sensors as well. Rebooting the device, the malfunction was oddly fixed 🤷‍♂️ 

So I thought it was just a strange bug in either Android or those devices.

Did you try to reboot your device? Does this fix the problem? Do other apps that read the accelerometer work as expected?

**Comment by glokuosta** [comment]
Yes, I have tried rebooting and even closing and reopening the device. I downloaded a sensor tool from Google Play and sensors work on that app with the same settings.

As I have said before, sensors worked on v2.6.1. I downloaded old versions of this app I have found on the internet, and the same shader works on 2.9.1 but not on 2.11. I couldn't find any built apk between those versions and I unfortunately don't have enough time to build them from source this year.

**Comment by joshcamas** [comment]
I'm having the same issue, also a Samsung :(

**Comment by glokuosta** [comment]
I did a little bit of homework and found that the breaking change is made in commit ae35d2c. Versions before that works normally.

**Comment by markusfisch** [comment]
Hm, that's weird. Obviously, [TYPE_LINEAR_ACCELERATION](https://developer.android.com/reference/android/hardware/Sensor.html#TYPE_LINEAR_ACCELERATION) and [TYPE_GRAVITY](https://developer.android.com/reference/android/hardware/Sensor.html#TYPE_GRAVITY) aren't available on your device, but [SensorManager.registerListener](https://developer.android.com/reference/android/hardware/SensorManager.html#registerListener(android.hardware.SensorEventListener,%20android.hardware.Sensor,%20int)) seems to return `true` 🤔 

The docs say, `registerListener` would return ...
> true if the sensor is supported and successfully enabled.

`TYPE_LINEAR_ACCELERATION` and `TYPE_GRAVITY` were added with API level 9. Before that, there was only [TYPE_ACCELEROMETER](https://developer.android.com/reference/android/hardware/Sensor.html#TYPE_ACCELEROMETER) and you had to calculate linear acceleration and gravity yourself.

A couple of days after the commit you found https://github.com/markusfisch/ShaderEditor/commit/ae35d2c8592b6c445b02235f8bc6a3ab84eda7b6, I've got a message from someone using a Galaxy J2 saying that the gravity sensor would no longer work. So I made the app use this new sensors (`TYPE_LINEAR_ACCELERATION` and `TYPE_GRAVITY` ) only if they can actually be registered in https://github.com/markusfisch/ShaderEditor/commit/257c5f9be078f0c6d61f3043c37d41b11886b790

And that's how things still are: https://github.com/markusfisch/ShaderEditor/blob/f856068097ca598916f3b1531b328899b0454445/app/src/main/java/de/markusfisch/android/shadereditor/opengl/ShaderRenderer.java#L935

So, since you've verified that `TYPE_ACCELEROMETER` works on your device, but the latest version of the app that uses `registerListener` to decide what sensor to use does not, it looks like the manufacturer of your device made `registerListener` return true when it shouldn't 😬

**Comment by glokuosta** [comment]
So I had some free time, and I created an Android Studio project with the code you linked in the android developer page, and the register function returned false. I also added some Log calls to the 935th line in the linked code and that returns false as well.

Then I added another log call to log the result of gravityValues at line 937 and logcat showed [0.0, 0.0, 0.0] as the result.

So, it seems that the function returns false, but the returned values from AccelerometerListener is wrong, although I don't know if they are supposed to be zero at first call.

**Comment by markusfisch** [comment]
You're right! I simply forgot to register the fallback accelerometer sensor in case `TYPE_LINEAR_ACCELERATION` and `TYPE_GRAVITY` fail to register 🙈

Fixed in  https://github.com/markusfisch/ShaderEditor/commit/b2835c5f32619164253a31c778a0a0c8f1a97d8c

But that just fixes the `gravity` and `linear` uniforms on devices without `TYPE_LINEAR_ACCELERATION` and `TYPE_GRAVITY`. `orientation` is calculated from the rotation matrix, what requires a magnetic field sensor to be calculated. Now, I found out that some of my test devices don't have `TYPE_MAGNETIC_FIELD` either.

Fortunately, we can also approximate the rotation matrix from the rotation vector with [SensorManager.getRotationMatrixFromVector](https://developer.android.com/reference/android/hardware/SensorManager.html#getRotationMatrixFromVector(float[],%20float[])). So I added a uniform for the rotation vector (https://github.com/markusfisch/ShaderEditor/commit/ec60e66b6fceebc90709ea94f4c10d198cc4e2e6) and use that to calculate the rotation matrix (and thus the device orientation) on devices without `TYPE_MAGNETIC_FIELD` (https://github.com/markusfisch/ShaderEditor/commit/49511f87a235a7a5391414896ae84684e33db1b0).

Just to find out, that my test device does not have a `TYPE_ROTATION_VECTOR` sensor too 🙈Maybe `TYPE_ROTATION_VECTOR` requires a magnetic field sensor in which case the fallback would be useless.

Anyway, I just built a new version, 2.16.0, with said changes and put it in beta so you can try it on your device:
https://play.google.com/apps/testing/de.markusfisch.android.shadereditor

**Comment by glokuosta** [comment]
So the gravity works, but orientation still doesn't work...

**Comment by ssartell** [comment]
Orientation.x seems to be the problem for me on my Pixel 3. If I take the sample orientation shader and switch it from orientation.z to orientation.x, it gets wonky. Both y and z seem to function as expected.

**Comment by Mehran7kl** [comment]
@markusfisch Orientation example doesn't work for me. My device works fine with games that use orientation for driving control. Can you fix it ?
